### PR TITLE
ffmpeg: update to 4.2.2

### PR DESCRIPTION
**Pull request by antonlacon**
Maintainer: me & @thess 
Compile tested: arm/mvebu git HEAD as of 2020-03-31
Run tested: arm/mvebu git HEAD as of 2020-03-31

Description:
Updates ffmpeg to 4.2.2. ffmpeg changes abi/api, but has been out hopefully long enough that there will be no fallout for this change.

ffserver is no longer exists. I'm not aware of a replacement.

Libavresample has been deprecated for a number of years and is no longer built by default, so don't include it here either. Programs should use libswresample instead.

Adds gnutls as the crypto library for libffmpeg-full. This should allows access to https streams, for example.

Tested by:
libffmpeg-full tested via ffmpeg cli tool (flac -> libopus and flac -> libfdk_aac); and
Rescan of minidlna media library.

### CODE REVIEW

**Comment by diizzyy** [comment]
I've made a new "one fits all" approach as the variants are causing issues with multiple packages here: https://github.com/openwrt/packages/pull/9475
Unfortunately it suffers the same recursive issue as the current ffmpeg port in tree.

**Comment by antonlacon** [comment]
The issues with the multiple packages are that they like to select libffmpeg-audio-dec?

File: multimedia/ffmpeg/Makefile [hunk 1]
```diff
@@ -345,7 +345,7 @@ endef
 define Package/libffmpeg-full
 $(call Package/libffmpeg/Default)
  TITLE+= (full)
- DEPENDS+= +alsa-lib +PACKAGE_libopus:libopus \
+ DEPENDS+= +alsa-lib +libgnutls +PACKAGE_libopus:libopus \
```
Comment (neheb) [inline multimedia/ffmpeg/Makefile:348]: Does ffmpeg actually support mbedtls?

Comment (diizzyy) [reply multimedia/ffmpeg/Makefile:348]: Yes, it does

**Comment by diizzyy** [comment]
You have issues where packages selects libffmpeg but support for various file types and codecs are missing etc making it quite a bit of a mess, minidlna comes to mind for instance.

**Comment by neheb** [comment]
minidlna's replacement seems to not use ffmpeg. It's really big though (1.8MB).

File: multimedia/ffmpeg/Makefile (hunk 1 above)
Comment (neheb) [reply multimedia/ffmpeg/Makefile:348]: I would think it's better to use that. It's smaller.

Comment (neheb) [reply multimedia/ffmpeg/Makefile:348]: hrm OK.

**Comment by neheb** [comment]
Here's a patch for the MIPS ASM failures: https://github.com/neheb/FFmpeg/commit/1317e256d884ff72a632a791a2dc64a9b1c3ef18

**Comment by neheb** [comment]
With ffmpeg-full I'm getting these configure errors:

```
WARNING: Option --disable-encoder=atrac3 did not match anything
WARNING: Option --disable-encoder=h264 did not match anything
WARNING: Option --disable-encoder=hevc did not match anything
WARNING: Option --disable-encoder=vc1 did not match anything
```

**Comment by diizzyy** [comment]
I'm pretty sure you'll need ffmpeg to make it somewhat usable in the end and most ARM-devices (or better) can at least transcode audio if needed which you probably want.
http://docs.gerbera.io/en/latest/compile.html

**Comment by neheb** [comment]
It says it's only for file metadata. Anyway, wrong PR.

**Comment by neheb** [comment]
MIPS ASM patch was merged upstream http://git.videolan.org/?p=ffmpeg.git;a=commit;h=875ba2333340c2e13af370832e49c9371ffb0f91

**Comment by neheb** [comment]
I noticed with ffmpeg-full after removing the disable ASM configuration option that h264 does not build under MIPS. Placing PKG_USE_MIPS16:=0 makes it compile again.

**Comment by antonlacon** [comment]
Thanks. I'll confirm it builds for me on mips and then push changes. I'll review the configure output on libffmpeg-full too. I'm guessing it's trying to disable them based on the patent flag.

Regarding the variant selection, my guess it would need some variant along the lines of libffmpeg-a[a-t]* to get it picked before libffmpeg-audio-dec that did what most people expected. I could see libffmpeg-audio-dec being renamed to libffmpeg-sound, but I'm not sure if libffmpeg-custom will get picked instead. libffmpeg-full could be renamed libffmpeg-all, but that might give people a surprise 6MB package; granted it would do everything.

**Comment by neheb** [comment]
This is my solution: https://github.com/neheb/packages/commit/e46bc7656304b4bc1f92ae1eed66a37c773c18a4

I tested all combinations to make sure they work.

**Comment by neheb** [comment]
ping@antonlacon

**Comment by antonlacon** [comment]
Thanks for ping.

Your mips related changes work for me. My previous target (ar71xx/generic/buffalo wzr-hp-g300nh) would fail before and compiles now.

The warnings you pointed out regarding the patented encoders, I've just dropped the calls to disable them entirely. None of them exist as named - atrac3 and vc1 don't exist at all to my knowledge, and h264 and hevc are done as other libraries. Of those, only libx264 is in openwrt, and ffmpeg relies on it being selected in the config before building support for it.

**Pull request merged**